### ISSUE TITLE
Fortran asks questions about the fibonacci benchmark

### ISSUE THREAD

**Issue by PEZ** [bddicken/languages#296]
Check these **fibonacci** benchmark results out:

<details><summary>Fortran seems to do it in O(1), right?</summary>

```
Checking C
Check passed
Benchmarking C
Benchmark 1: ./c/code 40
  Time (mean ± σ):     256.3 ms ±   6.7 ms    [User: 251.9 ms, System: 3.0 ms]
  Range (min … max):   249.0 ms … 262.2 ms    3 runs
 

Checking Clojure
Check passed
Benchmarking Clojure
Benchmark 1: java -cp clojure/classes:src:/Users/pez/.m2/repository/org/c ...
  Time (mean ± σ):     847.3 ms ±   8.5 ms    [User: 1082.1 ms, System: 59.1 ms]
  Range (min … max):   837.6 ms … 853.3 ms    3 runs
 

Checking Clojure Native
Check passed
Benchmarking Clojure Native
Benchmark 1: ./clojure-native-image/code 40
  Time (mean ± σ):     296.1 ms ±   5.5 ms    [User: 290.2 ms, System: 4.3 ms]
  Range (min … max):   292.1 ms … 302.5 ms    3 runs
 

Checking CPP
Check passed
Benchmarking CPP
Benchmark 1: ./cpp/code 40
  Time (mean ± σ):     264.3 ms ±   2.1 ms    [User: 260.5 ms, System: 2.8 ms]
  Range (min … max):   262.8 ms … 266.7 ms    3 runs
 

No executable or script found for Dart. Skipping.

Checking Fortran
Check passed
Benchmarking Fortran
Benchmark 1: ./fortran/code 40
  Time (mean ± σ):       4.5 ms ±   0.5 ms    [User: 2.1 ms, System: 2.0 ms]
  Range (min … max):     4.0 ms …   4.9 ms    3 runs
 

Checking Java
Check passed
Benchmarking Java
Benchmark 1: java jvm.code 40
  Time (mean ± σ):     339.2 ms ±  16.7 ms    [User: 329.9 ms, System: 31.5 ms]
  Range (min … max):   329.0 ms … 358.5 ms    3 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 

Checking Java GraalVM
Check passed
Benchmarking Java GraalVM
Benchmark 1: ./jvm.code 40
  Time (mean ± σ):     271.8 ms ±   1.4 ms    [User: 266.6 ms, System: 3.8 ms]
  Range (min … max):   270.4 ms … 273.1 ms    3 runs
 

Checking Node
Check passed
Benchmarking Node
Benchmark 1: node ./js/code.js 40
  Time (mean ± σ):     872.0 ms ±   7.4 ms    [User: 857.9 ms, System: 9.6 ms]
  Range (min … max):   867.6 ms … 880.6 ms    3 runs
 
  Warning: The first benchmarking run for this command was significantly slower than the rest (880.6 ms). This could be caused by (filesystem) caches that were not filled until after the first run. You are already using the '--warmup' option which helps to fill these caches before the actual benchmark. You can either try to increase the warmup count further or re-run this benchmark on a quiet system in case it was a random outlier. Alternatively, consider using the '--prepare' option to clear the caches before each timing run.
```
</details>

Looking at the Fortran code, without knowing any Fortran, it looks fine:

```fortran
program main
    implicit none
    integer(kind=4) :: u, r, i
    character(len=256) :: arg

    call get_command_argument(1, arg)
    read(arg, *) u

    r = 0

    do i = 1, u - 1
        r = r + fibonacci(i)
    end do

    print *, r

contains

    recursive function fibonacci(n) result(f)
        integer(kind=4), intent(in) :: n
        integer(kind=4) :: f

        if (n == 0) then
            f = 0
        elseif (n == 1) then
            f = 1
        else
            f = fibonacci(n - 1) + fibonacci(n - 2)
        end if
    end function fibonacci

end program main
```

But maybe that `recursive` directive tells the compiler to do tail recursion?

**Comment by PEZ** [comment]
@bddicken

**Comment by PEZ** [comment]
From what I can tell this is mainly showing that the Fortran compiler wins against the rest in figuring out that it can apply tail elimination.

**Comment by bddicken** [comment]
I think you had mentioned that you figured this out, is that right @PEZ? Or is fortran still behaving in this way?

**Comment by PEZ** [comment]
Fortran keeps behaving this way, but I think it is legit. I'm no Casey Muratori, so I may not be analyzing it correctly, but anyway, here's where I'm at as of current:

* https://github.com/PEZ/languages-visualizations/issues/2

TL;DR:

* Fortran finds some interesting optimizations for low fibonacci numbers that makes it crazy fast at `fib(40)`.
* The optimizations seem to enter in `gfortran` versions above 11 (I'm using version 14 something)
* At higher fibonacci numbers it starts to behave at exponential complexity, just as all other programs do
* To be clear: Fortran does not make it into tail eliminated recursion, as I suspected at first
* _All programs I have tested so far, except Clojure's, start to produce garbage answers at `fib(48)`_.

Here's a snapshot of the Fibonacci results from my latest run: https://pez.github.io/languages-visualizations/v2025.01.11/#fibonacci

This is a ***winner snapshot*** from that run:

![languages-visualizations-fibonacci-2025-01-11-gcc](https://github.com/user-attachments/assets/abc344d5-9ab2-4c59-8da4-11dd8eb6eab5)

Closing this issue for now. We can reopen if we find reasons to keep questioning the results.
